### ISSUE TITLE
Failure during streaming after the new node was added to the new data center

### ISSUE THREAD

**Issue by juliayakovlev** [scylladb/scylla#4347]
*Installation details*
Scylla version (or git commit hash): master( 64fa2dd)
Cluster size: 4 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Fedora

I ran dtest materialized_views.py TestMaterializedViews.add_dc_during_mv_insert_test.
Scenario:
- start cluster with 4 nodes in data center 1
- create the table with materialized view (RF=3)
- wait for view building
- prefill with data
- add new node to data center 2 in parallel with inserts into the table/MV

During streaming coredump was received:
```
INFO  2019-03-20 07:43:51,918 [shard 1] stream_session - [Stream #243133f0-4ad3-11e9-9bcb-000000000001] Write to sstable for ks=ks, cf=tm_table, estimated_partitions=3, received_partitions=248
INFO  2019-03-20 07:43:51,940 [shard 0] stream_session - [Stream #243133f0-4ad3-11e9-9bcb-000000000001] Write to sstable for ks=ks, cf=tm_table, estimated_partitions=3, received_partitions=282
scylla: read_context.hh:108: utils::phased_barrier::phase_type cache::autoupdating_underlying_reader::creation_phase() const: Assertion `_reader' failed.
Aborting on shard 0.
Backtrace:
  0x0000000004254382
  0x0000000004151835
  0x0000000004151b35
  0x0000000004151be3
  /lib64/libpthread.so.0+0x0000000000012aef
  /lib64/libc.so.6+0x000000000003766a
  /lib64/libc.so.6+0x0000000000039380
  /lib64/libc.so.6+0x000000000002f8f9
  /lib64/libc.so.6+0x000000000002f971
  0x0000000000fee337
  0x000000000414ecc1
  0x000000000414eebe
  0x0000000004224275
  0x00000000040ba205
  0x00000000009f6204
  /lib64/libc.so.6+0x0000000000021009
  0x0000000000a55eed
```
Decoded:
```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/boost/program_options/variables_map.hpp:146
seastar::backtrace_buffer::append_backtrace() at /usr/include/boost/program_options/variables_map.hpp:146
 (inlined by) print_with_backtrace at /home/juliay/scylla_repo/scylla/seastar/build/release/../../src/core/reactor.cc:1096
seastar::print_with_backtrace(char const*) at /usr/include/boost/program_options/variables_map.hpp:146
sigabrt_action at /usr/include/boost/program_options/variables_map.hpp:146
 (inlined by) operator() at /home/juliay/scylla_repo/scylla/seastar/build/release/../../src/core/reactor.cc:4911
 (inlined by) _FUN at /home/juliay/scylla_repo/scylla/seastar/build/release/../../src/core/reactor.cc:4907
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
seastar::continuation<seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > seastar::future<seastar::optimized_optional<mutation_fragment> >::then_impl<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > >(range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&)::{lambda(auto:1&&)#1}, seastar::optimized_optional<mutation_fragment> >::run_and_dispose() at crtstuff.c:?
 (inlined by) range_populating_reader::can_set_continuity() const at /home/juliay/scylla_repo/scylla/row_cache.cc:488
 (inlined by) range_populating_reader::can_set_continuity() const at /home/juliay/scylla_repo/scylla/row_cache.cc:487
 (inlined by) range_populating_reader::handle_end_of_stream() at /home/juliay/scylla_repo/scylla/row_cache.cc:491
 (inlined by) _ZZN23range_populating_readerclENSt6chrono10time_pointIN7seastar12lowres_clockENS0_8durationIlSt5ratioILl1ELl1000EEEEEEENUlOT_E_clINS2_18optimized_optionalI17mutation_fragmentEEEEDaSA_ at /home/juliay/scylla_repo/scylla/row_cache.cc:528
 (inlined by) seastar::apply_helper<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, std::tuple<seastar::optimized_optional<mutation_fragment> >&&, std::integer_sequence<unsigned long, 0ul> >::apply({lambda(auto:1&&)#1}&&, std::tuple<seastar::optimized_optional<mutation_fragment> >) at /home/juliay/scylla_repo/scylla/seastar/include/seastar/core/apply.hh:35
 (inlined by) auto seastar::apply<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, seastar::optimized_optional<mutation_fragment> >(range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&, std::tuple<seastar::optimized_optional<mutation_fragment> >&&) at /home/juliay/scylla_repo/scylla/seastar/include/seastar/core/apply.hh:43
 (inlined by) seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > seastar::futurize<seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > >::apply<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, seastar::optimized_optional<mutation_fragment> >(range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&, std::tuple<seastar::optimized_optional<mutation_fragment> >&&) at /home/juliay/scylla_repo/scylla/seastar/include/seastar/core/future.hh:1478
 (inlined by) seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > seastar::future<seastar::optimized_optional<mutation_fragment> >::then_impl<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > >(range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&)::{lambda(auto:1&&)#1}::operator()<seastar::future_state<seastar::optimized_optional<mutation_fragment> > >(auto, range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&) at /home/juliay/scylla_repo/scylla/seastar/include/seastar/core/future.hh:1015
 (inlined by) seastar::continuation<seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > seastar::future<seastar::optimized_optional<mutation_fragment> >::then_impl<range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}, seastar::future<seastar::optimized_optional<flat_mutation_reader>, seastar::optimized_optional<mutation_fragment> > >(range_populating_reader::operator()(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(auto:1&&)#1}&&)::{lambda(auto:1&&)#1}, seastar::optimized_optional<mutation_fragment> >::run_and_dispose() at /home/juliay/scylla_repo/scylla/seastar/include/seastar/core/future.hh:425
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/include/boost/program_options/variables_map.hpp:146
seastar::reactor::run_some_tasks() at /usr/include/boost/program_options/variables_map.hpp:146
seastar::reactor::run_some_tasks() at /usr/include/boost/program_options/variables_map.hpp:146
 (inlined by) seastar::reactor::run() at /home/juliay/scylla_repo/scylla/seastar/build/release/../../src/core/reactor.cc:4149
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/include/c++/8/bits/basic_string.h:312
main at crtstuff.c:?
?? ??:0
_start at ??:?
```
Scylla binary and core file are here: http://upload.scylladb.com/8d861b80-f94b-4994-90c4-7a06b62edb84

**Comment by juliayakovlev** [comment]
Same failure happens in the decommission test, where during node2 decommission the streaming on node4 failed

**Comment by tgrabiec** [comment]
Probably a duplicate of https://github.com/scylladb/scylla/issues/4236
